crates/solver/src/settlement/settlement_encoder.rs

```diff
@@ -343,81 +349,39 @@ impl SettlementEncoder {
-            .context("custom_buy_price computation failed")
-    }
-
-    /// Uses either the existing UCP for the sell token or the sell price within the order to
-    /// compute the price of the buy token right at the limit price of the passed order.
-    /// This is needed to settle liquidity orders in a way that gives them no surplus.
-    fn compute_limit_buy_price(&self, order: &Order) -> Result<U256> {
-        // Liquidity orders are settled at their limit price. We set:
-        // buy_price = sell_price * order.sellAmount / order.buyAmount, where sell_price is given from uniform clearing prices
-
-        // Rounding error checks:
-        // Following limit price constraint is checked in the smart contract:
-        // order.sellAmount.mul(sellPrice) >= order.buyAmount.mul(buyPrice),
-
-        // This equation will always be true, as the following equivalence is showing
-        // order.sellAmount.mul(sellPrice)  >= order.sellAmount.mul(sellPrice) holds, as the left and ride side are equal.
-        // Dividing first and then multiplying by order.buyAmount can only make the right side smaller
-        // <=> order.sellAmount.mul(sellPrice)  >= order.buyAmount.mul(sell_price * order.sellAmount / order.buyAmount)
-        // <=> order.sellAmount.mul(sellPrice)  >= order.buyAmount.mul(buyPrice)
-        // <=> equation from smart contract
-
-        self.clearing_prices
-            .get(&order.data.sell_token)
-            .unwrap_or(&order.data.buy_amount)
-            .checked_mul(order.data.sell_amount)
-            .context("buy_price calculation failed")?
-            .checked_div(order.data.buy_amount)
-            .context("buy_price calculation failed")
+        Ok((
+            uniform_sell_price,
+            sell_amount
+                .checked_mul(uniform_sell_price)
+                .context("custom_buy_price computation failed")?
+                .checked_div(buy_amount)
+                .context("custom_buy_price computation failed")?,
+        ))
     }
 
     fn add_custom_price_trade(
         &mut self,
         order: Order,
         executed_amount: U256,
         scaled_unsubsidized_fee: U256,
+        sell_price: U256,
         buy_price: U256,
     ) -> Result<TradeExecution> {
         verify_executed_amount(&order, executed_amount)?;
-        // For the encoding strategy of liquidity and limit orders, the sell prices are taken from
-        // the uniform clearing price vector. Therefore, either there needs to be an existing price
-        // for the sell token in the uniform clearing prices or we have to create a new price entry beforehand,
-        // if the sell token price is not yet available
-        if self.token_index(order.data.sell_token).is_none() {
-            let sell_token = order.data.sell_token;
-            let sell_price = order.data.buy_amount;
-            self.tokens.push(sell_token);
-            self.clearing_prices.insert(sell_token, sell_price);
```

Comment:
For context the assumption is that this line added the `buy_amount` of a foreign liquidity order contained in a solution to the clearing price vector of the `Settlement` if that order's `sell_token` was not already part of the clearing prices. Because this price is not denominated in ETH and actually only makes sense in relation to the custom price associated to that order it can make the `Settlement` fail the price deviation check.